### PR TITLE
feat(formatter): rustfmt page expression islands

### PR DESCRIPTION
**Pull request by kent8192**
## Summary

This PR addresses:

- Adds rustfmt formatting for safe Rust expression islands inside `page!` DSL macros.
- Extends the Page DSL grammar with semantic interpolation/path parsing needed for formatter-safe Rust islands.
- Applies the new formatter output to affected page/admin fixtures and updates the matching trybuild snapshot.

## Type of Change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactoring (code change that neither fixes a bug nor adds a feature)
- [ ] Documentation update
- [ ] Performance improvement
- [ ] Code quality improvements
- [ ] CI/CD changes
- [ ] Other (please describe):

## Breaking Change Assessment

- [x] **No** - This change is backward compatible
- [ ] **Yes** - This change breaks existing public API or behavior

## Motivation and Context

- The formatter previously delegated only surrounding Rust source to rustfmt, leaving Rust expressions inside `page!` DSL macros untouched.
- Formatting safe expression islands improves consistency while preserving unsupported, comment-bearing, and statement-style islands unchanged.

Refs #4841

Related to: #

## How Was This Tested?

- `cargo test -p tree-sitter-reinhardt-page -- --nocapture`
- `cargo test -p reinhardt-formatter -- --nocapture`
- `cargo clippy -p reinhardt-formatter --all-targets -- -D warnings`
- `cargo make fmt-check`
- `cargo test -p reinhardt-pages --test ui -- --nocapture`
- `git diff --check`

## Performance Impact

- No benchmarked performance impact. The new rustfmt pass only runs for detected safe `page!` Rust expression islands during explicit formatter execution.

## Checklist

- [x] I have followed the [Contributing Guidelines](../blob/main/CONTRIBUTING.md)
- [x] I have followed the [Commit Guidelines](../blob/main/instructions/COMMIT_GUIDELINE.md)
- [x] I have updated the documentation (if applicable)
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] I have tested with all affected database backends (if applicable)
- [x] I have formatted the code with `cargo make fmt-fix`
- [ ] I have checked the code with `cargo make clippy-check`
- [ ] I use self-hosted runner for CI (Repository owner only)

## Related Issues

- Refs #4841

### Type Label (select one)
- [ ] `bug` - Bug fix
- [x] `enhancement` - New feature or improvement
- [ ] `documentation` - Documentation update
- [ ] `performance` - Performance improvement
- [ ] `refactoring` - Code refactoring
- [ ] `code-quality` - Code quality improvements

### Additional Labels (apply alongside type label when applicable)
- [ ] `breaking-change` - Breaking change (**MUST** match "Yes" in Breaking Change Assessment above)

### Scope Label (select all that apply)
- [ ] `database` - Database layer, schema, migrations
- [ ] `auth` - Authentication, authorization, sessions
- [ ] `orm` - ORM layer, models, query builder
- [ ] `http` - HTTP layer, handlers, middleware
- [ ] `routing` - URL routing, path matching
- [ ] `api` - REST API, serializers, views
- [ ] `admin` - Admin interface, admin panels
- [ ] `forms` - Form handling, validation, rendering
- [ ] `graphql` - GraphQL schema, resolvers
- [ ] `websockets` - WebSocket connections, handlers
- [ ] `i18n` - Internationalization, localization
- [ ] `ci-cd` - CI/CD workflow changes

### Priority Label (for maintainers)
- [ ] `critical` - Blocks release or major functionality
- [ ] `high` - Important fix or feature
- [ ] `medium` - Normal priority
- [ ] `low` - Minor fix or enhancement

---

**Additional Context:**

- Comment-bearing and statement-style Rust islands are preserved byte-for-byte to avoid changing comments, raw strings, or multi-statement interpolation semantics.

## Summary by CodeRabbit

* **New Features**
  * Formatter now supports formatting Rust expression islands within page DSL macros using rustfmt, while preserving unsupported code regions.

* **Refactor**
  * Enhanced page grammar to better recognize closures, control flow, attributes, and events as distinct DSL constructs.
  * Refactored formatter engine to handle rustfmt options and improve code organization.

* **Tests**
  * Added comprehensive grammar parsing tests and rustfmt island formatting edge-case coverage.

### CODE REVIEW

**Comment by kent8192** [comment]
@codex review thoroughtly

File: crates/reinhardt-formatter/src/format_engine.rs [hunk 1]
```diff
@@ -334,116 +393,425 @@ fn format_dsl(kind: MacroKind, input: &str) -> Result<String, String> {
...
+	line_indent: &str,
+	rustfmt_options: &RustfmtOptions,
+) -> Option<String> {
+	if topiary_sensitive_rustfmt_island(island) {
+		return None;
+	}
+	let wrapped = format!("fn main() {{ let __reinhardt_fmt = {island}; }}\n");
+	let mut cmd = Command::new("rustfmt");
+	rustfmt_options.apply_to_command(&mut cmd);
+	if rustfmt_options.config_path.is_none() && rustfmt_options.edition.is_none() {
```
Comment (kent8192) [inline crates/reinhardt-formatter/src/format_engine.rs]: The island rustfmt pass should preserve the Rust 2024 requirement even when a config file is supplied. `fmt` / `fmt-all` auto-fill `config_path`, so this branch skips the fallback `--edition=2024`; an explicit config without `edition = "2024"` would format islands with rustfmt's default edition. Please default the island rustfmt invocation to edition 2024 whenever `edition` is not explicitly set, or otherwise ensure `RustfmtOptions` always carries that edition.

---
Reviewed-by: Codex

File: crates/reinhardt-formatter/src/format_engine.rs [hunk 2]
```diff
@@ -334,116 +393,425 @@ fn format_dsl(kind: MacroKind, input: &str) -> Result<String, String> {
...
-		})
+	let formatted = String::from_utf8(output).map_err(|e| {
+		format!(
+			"Topiary produced invalid UTF-8 for {} DSL: {e}",
+			kind.name()
+		)
+	})?;
+	let mut normalized = normalize_dsl_output(formatted.trim());
+	if kind == MacroKind::Page {
+		normalized = format_page_rustfmt_islands(&normalized, rustfmt_options);
```
Comment (kent8192) [inline crates/reinhardt-formatter/src/format_engine.rs:411]: This gate makes the rustfmt-island pass `page!`-only. That matches the PR description, but the linked issue also covers `form!` and `head!`. Please either expand coverage here or update the PR / issue linkage from `Fixes #4841` to a partial reference so the remaining macro coverage stays tracked.

---
Reviewed-by: Codex

File: crates/reinhardt-formatter/src/format_engine.rs [hunk 3]
```diff
@@ -334,116 +393,425 @@ fn format_dsl(kind: MacroKind, input: &str) -> Result<String, String> {
...
+
+fn restore_protected_rustfmt_islands(input: &mut String, islands: &[ProtectedRustfmtIsland]) {
+	for island in islands {
+		*input = input.replace(&island.marker, &island.original);
+	}
 }
 
 fn normalize_dsl_output(input: &str) -> String {
-	normalize_unary_minus(input)
+	input.to_string()
```
Comment (kent8192) [inline crates/reinhardt-formatter/src/format_engine.rs:426]: Dropping the shared unary-minus normalizer changes the non-page path because only `page!` now gets rustfmt islands. Please add explicit `form!` / `head!` regression tests for cases like `(- 1_i32)`, or restore equivalent normalization for those macros.

---
Reviewed-by: Codex

File: crates/reinhardt-admin-cli/README.md [hunk 4]
```diff
@@ -129,7 +129,9 @@ reinhardt-admin plugin update --all
 
 Use `reinhardt-formatter` to format all Rust files in the project. Reinhardt
 DSL macros are formatted with tree-sitter grammars and Topiary queries before
-`cargo fmt --all` formats the surrounding Rust code:
+`cargo fmt --all` formats the surrounding Rust code. For supported `page!`
```
Comment (kent8192) [inline crates/reinhardt-admin-cli/README.md:132]: This documents a new formatter behavior, and the `--with-rustfmt=false` semantics are clarified in the CLI help. Please add a matching Unreleased entry in `crates/reinhardt-formatter/CHANGELOG.md` so the user-visible formatter change is captured for release notes.

---
Reviewed-by: Codex

File: crates/tree-sitter-reinhardt-page/src/scanner.c [hunk 5]
```diff
@@ -38,6 +45,399 @@ static bool is_ident_continue(int32_t c) {
...
+	bool has_significant_tail = false;
+	int32_t last_significant = 0;
+	char last_identifier[16];
+	unsigned last_identifier_len = 0;
+	bool identifier_active = false;
+
+	while (lexer->lookahead != 0) {
+		if (lexer->lookahead == '{' && paren_depth == 0 && bracket_depth == 0) {
+			if (control_head_block_starts_expression(token_type, has_significant_tail, last_significant,
+													 last_identifier, last_identifier_len)) {
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/tree-sitter-reinhardt-page/src/scanner.c]: Handled in 7b7ddff by tracking if-let and for-loop pattern context in the scanner so struct-pattern brace blocks remain part of the control-flow head. Added parser and formatter regression coverage for `if let Foo { ... }` and `for Foo { ... }`.

🤖 Generated with [Codex](https://openai.com/codex)

File: crates/tree-sitter-reinhardt-page/grammar.js [hunk 6]
```diff
@@ -27,10 +49,97 @@ module.exports = grammar({
...
+      '{',
+      $.rustfmt_island,
+      '}',
+    )),
+
     block: $ => seq('{', repeat($._item), '}'),
     paren: $ => seq('(', repeat($._item), ')'),
     bracket: $ => seq('[', repeat($._item), ']'),
 
+    rustfmt_island: $ => prec.right(repeat1(choice($._rust_item, $.semicolon))),
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/tree-sitter-reinhardt-page/grammar.js:119]: Handled in 7b7ddff by tokenizing closure args inside page Rust islands and preserving Rust syntax commas for multi-parameter closures and turbofish arguments. Added parser and formatter regression coverage for both cases.

🤖 Generated with [Codex](https://openai.com/codex)

File: crates/reinhardt-formatter/src/format_engine.rs (hunk 1 above)
Comment (kent8192) [reply crates/reinhardt-formatter/src/format_engine.rs]: Handled in 7b7ddff by defaulting page Rust-island rustfmt invocations to `--edition=2024` whenever no edition is explicitly provided, including the `config_path` case. Added regression coverage with a config file path.

🤖 Generated with [Codex](https://openai.com/codex)

File: crates/reinhardt-formatter/src/format_engine.rs (hunk 2 above)
Comment (kent8192) [reply crates/reinhardt-formatter/src/format_engine.rs:411]: Handled by updating the PR body from `Fixes #4841` to `Refs #4841`, keeping this PR scoped to page Rust expression islands while leaving form/head coverage tracked separately.

🤖 Generated with [Codex](https://openai.com/codex)

File: crates/reinhardt-formatter/src/format_engine.rs (hunk 3 above)
Comment (kent8192) [reply crates/reinhardt-formatter/src/format_engine.rs:426]: Handled in 7b7ddff by adding explicit form/head regression coverage that preserves `tabindex: (-1_i32)` on the non-page formatting path.

🤖 Generated with [Codex](https://openai.com/codex)

File: crates/reinhardt-admin-cli/README.md (hunk 4 above)
Comment (kent8192) [reply crates/reinhardt-admin-cli/README.md:132]: Handled in 7b7ddff by adding an Unreleased entry to `crates/reinhardt-formatter/CHANGELOG.md` for the user-visible page Rust-island formatter behavior.

🤖 Generated with [Codex](https://openai.com/codex)
